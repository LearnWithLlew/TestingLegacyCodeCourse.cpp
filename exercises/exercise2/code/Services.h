#pragma once

#include <iostream>
#include <sstream>

#include "third_party_code.h"
#if ENABLE_FAKE_REST_CALL
std::stringstream logstream;
#endif

int RestCall(std::string url, std::string action, int key)
{
#if ENABLE_FAKE_REST_CALL
    logstream << "RestCall(" << url << "," << action << "," << key << ")\n";
    logstream << "RESULT: " << 200 << "\n";
    return 200;
#else
    return _third_party_code_::rest_call(url, action, key);
#endif
}

void initialiseServices()
{
    int key = 42;
    int result = RestCall("https://postgresdatabase.acme.com", "start", key);
    if ( result == 200 )
    {
        key = key + 100;
        result = RestCall("https://kafka.acme.com", "subscribe", key);
        if ( result == 200 )
        {
            result = RestCall("https://cache.acme.com", "load", key);
            if ( result != 200 )
            {
                std::cout << "Error starting cache\n";
            }
            result = RestCall("https://search.acme.com", "index", key);
            if ( result != 200 )
            {
                std::cout << "Error indexing search\n";
            }
            result = RestCall("https://creditcardprocessor.acme.com", "charge", key);
            if ( result != 200 )
            {
                std::cout << "Error charging credit card\n";
            }
        }
    }
}
