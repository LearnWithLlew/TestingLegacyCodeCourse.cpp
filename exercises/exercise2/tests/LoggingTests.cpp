#include <catch2/catch.hpp>
#include "ApprovalTests.hpp"

#define SECURE 1
#define ENABLE_FAKE_REST_CALL 1
#include "Services.h"

#include <iostream>

using namespace ApprovalTests;

TEST_CASE("Test initialiseServices2")
{
	logstream = std::stringstream{};
	initialiseServices();
	Approvals::verify(logstream.str());
}

TEST_CASE("Test initialiseServices")
{
	logstream = std::stringstream{};
	initialiseServices();
	Approvals::verify(logstream.str());
}
