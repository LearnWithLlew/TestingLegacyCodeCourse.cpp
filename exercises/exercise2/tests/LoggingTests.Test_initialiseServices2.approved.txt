RestCall(https://postgresdatabase.acme.com,start,42)
RESULT: 200
RestCall(https://kafka.acme.com,subscribe,142)
RESULT: 200
RestCall(https://cache.acme.com,load,142)
RESULT: 200
RestCall(https://search.acme.com,index,142)
RESULT: 200
RestCall(https://creditcardprocessor.acme.com,charge,142)
RESULT: 200

